### ISSUE TITLE
Feature: allow the specification of the value for the parser constants

### ISSUE THREAD

**Issue by zosrothko** [javacc/javacc#34]
Hi

For the C++ generation, JavaCC generates a ParserConstants file which is defining the token's kind as a list of  const int TOKEN = <integer> as for example:

With such TOKEN declaration:
```
TOKEN							:
{
  <#NA				: ~["-","\r","\n"] >
| <#EOL				: ("\n" | "\r" | "\r\n") >
}
TOKEN 				:
{
  <CTAG				: "/>">
| <ETAG				: "</">
| <LBRACKET 			: "["> 
| <RBRACKET 			: "]">
| <LBRACE 			: "{">
| <RBRACE 			: "}">
```
JavaCC generates:
```
/* Generated By:JavaCC: Do not edit this line. TyperConstants.java */
/**
 * Token literal values and constants.
 * Generated by org.javacc.parser.OtherFilesGen#start()
 */
#ifndef TYPERCONSTANTS_H
#define TYPERCONSTANTS_H
#include "JavaCC.h"

namespace ASN1 {
namespace Typer {
/** End of File. */
const  int _EOF = 0;
/** RegularExpression Id. */
const  int NA = 9;
/** RegularExpression Id. */
const  int EOL = 10;
/** RegularExpression Id. */
const  int CTAG = 13;
/** RegularExpression Id. */
const  int ETAG = 14;
/** RegularExpression Id. */
const  int LBRACKET = 15;

```
This issue is requesting to permit the customization of the integer value given to each token's const int as for example
TOKEN							:
{
  <#NA				: ~["-","\r","\n"] >           : USER_NA
| <#EOL				: ("\n" | "\r" | "\r\n") >    : USER_EOL
}
TOKEN 				:
{
  <CTAG				: "/>">                            : USER_CTAG
| <ETAG				: "</">                            : USER_ETAG
| <LBRACKET 			: "[">                              : USER_LBRACKET 			
| <RBRACKET 			: "]">                              : USER_RBRACKET 			
| <LBRACE 			: "{">                              : USER_LBRACE 			
| <RBRACE 			: "}">                              : USER_RBRACE 			
```
so that the generated constant file looks like
```
/* Generated By:JavaCC: Do not edit this line. TyperConstants.java */
/**
 * Token literal values and constants.
 * Generated by org.javacc.parser.OtherFilesGen#start()
 */
#ifndef TYPERCONSTANTS_H
#define TYPERCONSTANTS_H
#include "JavaCC.h"

namespace ASN1 {
namespace Typer {
/** End of File. */
const  int _EOF = 0;
/** RegularExpression Id. */
const  int NA = USER_NA;
/** RegularExpression Id. */
const  int EOL = USER_EOL ;
/** RegularExpression Id. */
const  int CTAG = USER_CTAG ;
/** RegularExpression Id. */
const  int ETAG = USER_ETAG ;
/** RegularExpression Id. */
const  int LBRACKET = USER_LBRACKET ;

```

**Comment by new-javacc** [comment]
I think this is unnecessary. You can just use the same name as the token name in the grammar. I think this is a gratuitous feature requestOn Nov 1, 2017 9:47 AM, zosrothko <notifications@github.com> wrote:Hi
For the C++ generation, JavaCC generates a ParserConstants file which is defining the token's kind as a list of  const int TOKEN =  as for example:
With such TOKEN declaration:
TOKEN							:
{
  <#NA				: ~["-","\r","\n"] >
| <#EOL				: ("\n" | "\r" | "\r\n") >
}
TOKEN 				:
{
  <CTAG				: "/>">
| <ETAG				: "</">
| <LBRACKET 			: "["> 
| <RBRACKET 			: "]">
| <LBRACE 			: "{">
| <RBRACE 			: "}">

JavaCC generates:
/* Generated By:JavaCC: Do not edit this line. TyperConstants.java */
/**
 * Token literal values and constants.
 * Generated by org.javacc.parser.OtherFilesGen#start()
 */
#ifndef TYPERCONSTANTS_H
#define TYPERCONSTANTS_H
#include "JavaCC.h"

namespace ASN1 {
namespace Typer {
/** End of File. */
const  int _EOF = 0;
/** RegularExpression Id. */
const  int NA = 9;
/** RegularExpression Id. */
const  int EOL = 10;
/** RegularExpression Id. */
const  int CTAG = 13;
/** RegularExpression Id. */
const  int ETAG = 14;
/** RegularExpression Id. */
const  int LBRACKET = 15;


This issue is requesting to permit the customization of the integer value given to each token's const int as for example
TOKEN							:
{
<#NA				: ~["-","\r","\n"] >           : USER_NA
| <#EOL				: ("\n" | "\r" | "\r\n") >    : USER_EOL
}
TOKEN 				:
{
<CTAG				: "/>">                            : USER_CTAG
| <ETAG				: "</">                            : USER_ETAG
| <LBRACKET 			: "[">                              : USER_LBRACKET
| <RBRACKET 			: "]">                              : USER_RBRACKET
| <LBRACE 			: "{">                              : USER_LBRACE
| <RBRACE 			: "}">                              : USER_RBRACE
so that the generated constant file looks like

/* Generated By:JavaCC: Do not edit this line. TyperConstants.java /
/*
Token literal values and constants.Generated by org.javacc.parser.OtherFilesGen#start()
*/
#ifndef TYPERCONSTANTS_H
#define TYPERCONSTANTS_H
#include "JavaCC.h"
namespace ASN1 {
namespace Typer {
/** End of File. /
const  int _EOF = 0;
/* RegularExpression Id. /
const  int NA = USER_NA;
/* RegularExpression Id. /
const  int EOL = USER_EOL ;
/* RegularExpression Id. /
const  int CTAG = USER_CTAG ;
/* RegularExpression Id. /
const  int ETAG = USER_ETAG ;
/* RegularExpression Id. */
const  int LBRACKET = USER_LBRACKET ;

**Comment by new-javacc** [comment]
Also with my pluggable/extensible codegen framework, yih cnabdo these things as add-ons so please keep javacc clean for now.FYI you will see some of my changes in the new branch csharo_codegenOn Nov 1, 2017 10:01 AM, JavaCC Support <support@javacc.org> wrote:I think this is unnecessary. You can just use the same name as the token name in the grammar. I think this is a gratuitous feature requestOn Nov 1, 2017 9:47 AM, zosrothko <notifications@github.com> wrote:Hi
For the C++ generation, JavaCC generates a ParserConstants file which is defining the token's kind as a list of  const int TOKEN =  as for example:
With such TOKEN declaration:
TOKEN							:
{
  <#NA				: ~["-","\r","\n"] >
| <#EOL				: ("\n" | "\r" | "\r\n") >
}
TOKEN 				:
{
  <CTAG				: "/>">
| <ETAG				: "</">
| <LBRACKET 			: "["> 
| <RBRACKET 			: "]">
| <LBRACE 			: "{">
| <RBRACE 			: "}">

JavaCC generates:
/* Generated By:JavaCC: Do not edit this line. TyperConstants.java */
/**
 * Token literal values and constants.
 * Generated by org.javacc.parser.OtherFilesGen#start()
 */
#ifndef TYPERCONSTANTS_H
#define TYPERCONSTANTS_H
#include "JavaCC.h"

namespace ASN1 {
namespace Typer {
/** End of File. */
const  int _EOF = 0;
/** RegularExpression Id. */
const  int NA = 9;
/** RegularExpression Id. */
const  int EOL = 10;
/** RegularExpression Id. */
const  int CTAG = 13;
/** RegularExpression Id. */
const  int ETAG = 14;
/** RegularExpression Id. */
const  int LBRACKET = 15;


This issue is requesting to permit the customization of the integer value given to each token's const int as for example
TOKEN							:
{
<#NA				: ~["-","\r","\n"] >           : USER_NA
| <#EOL				: ("\n" | "\r" | "\r\n") >    : USER_EOL
}
TOKEN 				:
{
<CTAG				: "/>">                            : USER_CTAG
| <ETAG				: "</">                            : USER_ETAG
| <LBRACKET 			: "[">                              : USER_LBRACKET
| <RBRACKET 			: "]">                              : USER_RBRACKET
| <LBRACE 			: "{">                              : USER_LBRACE
| <RBRACE 			: "}">                              : USER_RBRACE
so that the generated constant file looks like

/* Generated By:JavaCC: Do not edit this line. TyperConstants.java /
/*
Token literal values and constants.Generated by org.javacc.parser.OtherFilesGen#start()
*/
#ifndef TYPERCONSTANTS_H
#define TYPERCONSTANTS_H
#include "JavaCC.h"
namespace ASN1 {
namespace Typer {
/** End of File. /
const  int _EOF = 0;
/* RegularExpression Id. /
const  int NA = USER_NA;
/* RegularExpression Id. /
const  int EOL = USER_EOL ;
/* RegularExpression Id. /
const  int CTAG = USER_CTAG ;
/* RegularExpression Id. /
const  int ETAG = USER_ETAG ;
/* RegularExpression Id. */
const  int LBRACKET = USER_LBRACKET ;

**Comment by zosrothko** [comment]
Anyway, I can use a external routine for matching the token type. It was a way to avoid misaligned values

**Comment by new-javacc** [comment]
I'm still curious about the original issue you are trying to solve. So if you can describe it briefly, we can see if there is a better way to do it.On Nov 2, 2017 9:55 AM, zosrothko <notifications@github.com> wrote:Closed #34.
